### PR TITLE
Minimal changes to avoid unnecessary DB lock on read

### PR DESCRIPTION
**Pull request by atz**
With accompanying tests to prove it.

### CODE REVIEW

**Comment by cbeer** [comment]
I don't understand the current locking strategy, and I don't understand the refinement you propose.

I believe `#next_id` needs to be an atomic operation, locking the row while it reads, mints an id, and then writes the new serialized state into the database. The indirection confuses me, can you help me understand why it is necessary?

**Comment by atz** [comment]
@cbeer Did you follow the conversation in channel over the past day?

The refinement here is to prevent the `read` operation locking for no reason.  

The indirection became necessary, according to @mjgiarlo's changes, to support migration rake tasks where only "half" the operation in a given type is performed (read in one type, write in another).

**Comment by atz** [comment]
Previous log analysis showed:
```
D, [2017-02-22T20:44:55.545635 #24996] DEBUG -- : [8fff3e6e-5d9d-4f5f-9a2e-54d551d79ed6] [ActiveJob] [CreateWorkJob] [88be6a3c-560f-4bf0-b058-39840e887c03]   MinterState Load (88478.5ms)  SELECT  "minter_states".* FROM "minter_states" WHERE "minter_states"."namespace" = $1 AND "minter_states"."template" = $2 LIMIT $3 FOR UPDATE  [["namespace", "default"], ["template", ".reeddeeddk"], ["LIMIT", 1]]
D, [2017-02-22T20:44:55.549649 #24996] DEBUG -- : [8fff3e6e-5d9d-4f5f-9a2e-54d551d79ed6] [ActiveJob] [CreateWorkJob] [88be6a3c-560f-4bf0-b058-39840e887c03]   MinterState Load (1.5ms)  SELECT  "minter_states".* FROM "minter_states" WHERE "minter_states"."namespace" = $1 AND "minter_states"."template" = $2 LIMIT $3 FOR UPDATE  [["namespace", "default"], ["template", ".reeddeeddk"], ["LIMIT", 1]]
D, [2017-02-22T20:44:55.552335 #24996] DEBUG -- : [8fff3e6e-5d9d-4f5f-9a2e-54d551d79ed6] [ActiveJob] [CreateWorkJob] [88be6a3c-560f-4bf0-b058-39840e887c03]   MinterState Exists (1.1ms)  SELECT  1 AS one FROM "minter_states" WHERE "minter_states"."namespace" = $1 AND ("minter_states"."id" != $2) LIMIT $3  [["namespace", "default"], ["id", 1], ["LIMIT", 1]]
D, [2017-02-22T20:44:55.555401 #24996] DEBUG -- : [8fff3e6e-5d9d-4f5f-9a2e-54d551d79ed6] [ActiveJob] [CreateWorkJob] [88be6a3c-560f-4bf0-b058-39840e887c03]   SQL (1.8ms)  UPDATE "minter_states" SET "seq" = $1, "counters" = $2, "rand" = $3, "updated_at" = $4 WHERE "minter_states"."id" = $5  [["seq", 2467], ["counters", "[{\"value\":6,\"max\":24139284}, ...
```

Note the duplicate queries, both **SELECT ... FOR UPDATE**.  This minimal change will make the first query a regular `SELECT` non-locking.  

We should go further for transactions and make it just one locking query, but I submit this first piece to aid our ongoing ingest task performance.

File: lib/active_fedora/noid/minter/db.rb [hunk 1]
```diff
@@ -43,10 +43,13 @@ def next_id
...
         end
 
-        def instance
-          MinterState.lock.find_by!(
-            namespace: ActiveFedora::Noid.config.namespace,
-            template: ActiveFedora::Noid.config.template
+        # @param [Boolean] lock whether the returned object (row) should be locked for update or not
+        # @return [MinterState]
+        def instance(lock = false)
+          MinterState.find_by!(
```
Comment (cbeer) [inline lib/active_fedora/noid/minter/db.rb]: Where is this style of lock documented? I tried to follow the chain and never found it. The Rails documentation seems to suggest `#lock(bool)` as maybe a preferred alternative, although I don't know what the Rails version support on that is like?

Comment (atz) [inline lib/active_fedora/noid/minter/db.rb]: It is documented on `find`

**Comment by cbeer** [comment]
> The refinement here is to prevent the read operation locking for no reason.

I don't understand why we'd want to read data that we know is going to be stale and is of no use for minting new identifiers?

File: lib/active_fedora/noid/minter/db.rb (hunk 1 above)
Comment (atz) [inline lib/active_fedora/noid/minter/db.rb]: http://apidock.com/rails/ActiveRecord/Base/find/class

**Comment by atz** [comment]
I have my own reservations about the indirection and I agree it is problematic.  My initial implementation did not have separate read and write locks and did not query twice for the same row.  I think this is a symptom of attempting to get too DRY without respecting the constraints.  The motivation was to support migration rake tasks.

But the indirection is not under review in this PR.  For that see #40.  I have already proposed to go further to eliminate the duplicate query, subsequent to this PR.

**Comment by cbeer** [comment]
Sure, but I believe making sure we're locking at an appropriate time is in-scope for review. Previously, we were making the lock when we read the serialized state. Now we're deferring that until we go and write to the database, are we not? If we wait that long, isn't it possible for multiple minters to come along, mint an id, and then write conflicting state information to the database?

File: lib/active_fedora/noid/minter/db.rb (hunk 1 above)
Comment (cbeer) [inline lib/active_fedora/noid/minter/db.rb]: I don't see where the lock gets applied. Here's an example:

```
irb(main):001:0> User.lock.find(1)
  User Load (3.9ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2 FOR UPDATE  [["id", 1], ["LIMIT", 1]]
```

vs

```
irb(main):006:0> User.find_by!({id: 1}, lock: true)
  User Load (0.4ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
=
```

**Comment by atz** [comment]
We read twice, as the log shows.  Once during `read`, once during `write!` when each method calls `instance`.  A separate object (locked) is returned.  `read` throws away its locked object.  No update is ever made to it.  You are correct that we are presuming the second object matches the first (and I don't like that), but that has nothing to do with whether the first object was locked or not.  The first object is destroyed before the second is created.

**Comment by cbeer** [comment]
Sure, but `#write!` isn't called until after we've minted an identifier using the potentially stale state information, is it not?

**Comment by jcoyne** [comment]
What Chris is saying is that `find` with `lock:true` does not do `SELECT...FOR UPDATE`

**Comment by cbeer** [comment]
@jcoyne: I'm saying that as well, in a code comment thread. I'm also saying with this change, I don't have confidence we're locking at the appropriate times (and I don't know how much confidence I had in the old code, but at least it seemed to try to do lots of locking)

**Comment by atz** [comment]
@cbeer I agree the only thing that might be mitigating that is the transaction block.  

In my original implementation, the transaction operated on a single object and this divergence was not possible.

**Comment by cbeer** [comment]
I don't understand how the transaction block mitigates anything? My understanding is that the lock lifetime is scoped to the transaction, but you'd still need to acquire that lock before reading data if you want to prevent a bad identifier from being minted. 

Would it help if I replicate the failure scenario I'm concerned about?

File: lib/active_fedora/noid/minter/db.rb (hunk 1 above)
Comment (atz) [inline lib/active_fedora/noid/minter/db.rb]: Ah, sorry, you are correct.  Will rework that.

**Comment by atz** [comment]
It will only help if it is applicable to this PR and not #40.  Your complaints are with changes already made.

**Comment by atz** [comment]
If you can conceive of a way to test multiple competing DB sessions in rspec, that would be rad though.

**Comment by cbeer** [comment]
Here's the difference between the current behavior and the one in #67:

`master`:
master

`db_lock_on_read`:
new

**Comment by atz** [comment]
I'm having trouble interpreting the flashing gifs there, and I assume they predate the rebase.

**Pull request closed without merging**